packages/core/src/shared/utilities/map.ts

```diff
@@ -0,0 +1,98 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getLogger } from '../logger'
+
+/**
+ * Simply maps a Key to an Object. Everything is done in memory, allowing for atomic updates.
+ *
+ * There is a common use case for this class, so look to reuse this when possible.
+ * Benefits, instead of making your own:
+ * - A key does not have to be a string, {@link MemoryMap.asKey} enables this
+ * - Implements most of the boilerplate
+ */
+export abstract class MemoryMap<Key, Value = { [key: string]: any }> implements MapSync<Key, Value> {
```

Comment:
most data is in-memory, so it seems confusing to call that out. if we add a non-memory variant then its name could call that out, since it would be less common.

```suggestion
export abstract class NestedMap<Key, Value = { [key: string]: any }> implements MapSync<Key, Value> {
```